### PR TITLE
fix(expr): `array_length` and `cardinality` returns int32 as in PostgreSQL

### PR DESCRIPTION
**Pull request by xiangjinwu**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

Since this commit, the frontend will generate plans with `int32`.

For backward compatibility, the backend will run both `int32` and `int64` variants as requested by the query plan.

Note: Having both `cardinality(list) -> int32` and `cardinality(list) -> int64` in the function signature map means the type inferring logic cannot find a unique match. But we have them covered in `infer_type_for_special` to always return `int32`. Eventually it is still better to clean them up from the function signature map.

## Checklist For Contributors

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

### Types of user-facing changes

- SQL commands, functions, and operators

### Release note

`array_length` and `cardinality` returns int32 as in PostgreSQL

### CODE REVIEW

File: src/expr/src/vector_op/array_length.rs [hunk 1]
```diff
@@ -126,8 +127,8 @@ fn array_length(array: ListRef<'_>) -> Result<i64, ExprError> {
 /// statement error
 /// select array_length(array[null, array[2]], 2);
 /// ```
-#[function("array_length(list, int32) -> int64")]
-fn array_length_d(array: ListRef<'_>, d: i32) -> Result<Option<i64>, ExprError> {
+#[function("array_length(list, int32) -> int32")]
+fn array_length_of_dim(array: ListRef<'_>, d: i32) -> Result<Option<i32>, ExprError> {
```
Comment (xiangjinwu) [inline src/expr/src/vector_op/array_length.rs:131]: This variant is just added by #10197 and we do not need to worry about backward compatibly.

**Review by lmatz — APPROVED** [review]

**Review by kwannoel — APPROVED** [review]

**Review by xxchan — APPROVED** [review]
> Eventually it is still better to clean them up from the function signature map.

Do you have any plan or idea about how to clean it?

**Comment by xxchan** [comment]
I listed the non-unique signatures in a test.

**Comment by xiangjinwu** [comment]
> > Eventually it is still better to clean them up from the function signature map.
> 
> Do you have any plan or idea about how to clean it?

#10285

**Pull request merged**